Allow running the mold test suite

lapla-cogito opened the pull request:
By specifying the path to mold's test suite (`mold/test`) through the `external_test_suites` option, we can execute mold's test suite in `cargo test`. Future patches will automatically classify which outputs should be properly ignored, such as error messages.

As some preliminary statistics, the following tests currently pass as is:
<details><summary>click to expand</summary>

```
test/tls-gd-dlopen.sh
test/crel.sh
test/emit-relocs-dead-sections.sh
test/glibc-2.22-bug.sh
test/rpath.sh
test/dynamic-list4.sh
test/allow-multiple-definition.sh
test/arch-x86_64-gnu-retain.sh
test/lto-nostdlib.sh
test/trace.sh
test/arch-riscv64-variant-cc.sh
test/protected.sh
test/gnu-property.sh
test/whole-archive.sh
test/bno-symbolic.sh
test/gnu-warning.sh
test/tls-irregular-start-addr.sh
test/arch-x86_64-tls-gd-to-ie.sh
test/strip-debug.sh
test/dynamic-list3.sh
test/linker-script5.sh
test/arch-x86_64-z-ibtplt.sh
test/empty-version.sh
test/tls-small-alignment.sh
test/large-max-page-size.sh
test/tlsdesc-local-dynamic.sh
test/lto-gcc.sh
test/response-file2.sh
test/compress-debug-sections-zstd.sh
test/sysroot2.sh
test/disable-new-dtags.sh
test/arch-aarch64-variant-pcs.sh
test/dynamic-dt-debug.sh
test/export-from-exe.sh
test/dt-init.sh
test/version-script2.sh
test/ifunc-static.sh
test/dynamic-list-data.sh
test/arch-x86_64-z-text.sh
test/lto-llvm.sh
test/mcmodel-large.sh
test/version-script4.sh
test/version-script13.sh
test/gc-sections.sh
test/section-attributes.sh
test/link-order.sh
test/arch-s390x-got.sh
test/gnu-unique.sh
test/range-extension-thunk3.sh
test/version-script12.sh
test/plt-dso.sh
test/arch-x86_64-reloc-zero.sh
test/dynamic.sh
test/func-addr.sh
test/textrel2.sh
test/dt-needed.sh
test/emit-relocs-cpp.sh
test/ifunc-static-pie.sh
test/pltgot.sh
test/arch-x86_64-mergeable-records.sh
test/arch-x86_64-relax.sh
test/linker-script4.sh
test/arch-riscv64-attributes2.sh
test/tls-large-static-image.sh
test/ifunc-dso.sh
test/arch-x86_64-isa-level.sh
test/unresolved-symbols2.sh
test/arch-x86_64-tls-ld-mcmodel-large.sh
test/ifunc-dlopen.sh
test/arch-x86_64-incompatible-obj.sh
test/lto-archive.sh
test/symbol-version.sh
test/arch-x86_64-tls-large-tbss.sh
test/arch-x86_64-unique.sh
test/hidden-weak-undef.sh
test/arch-arm-range-extension-thunk.sh
test/main-in-dso.sh
test/hash-style-sysv.sh
test/tlsdesc-dlopen.sh
test/large-text.sh
test/tls-gd.sh
test/dynamic-list2.sh
test/arch-x86_64-init-array.sh
test/mergeable-strings.sh
test/hidden-archive.sh
test/compress-debug-sections.sh
test/demangle-rust.sh
test/large-alignment.sh
test/arch-x86_64-incompatible-libs-linker-script2.sh
test/lto-dso.sh
test/version-script3.sh
test/sysroot.sh
test/range-extension-thunk.sh
test/library.sh
test/audit.sh
test/push-pop-state.sh
test/soname.sh
test/z-pack-relative-relocs.sh
test/tlsdesc-import.sh
test/tlsdesc-static.sh
test/hello-dynamic.sh
test/arch-x86_64-z-ibt.sh
test/symbol-version-lto.sh
test/symbol-version4.sh
test/init-in-dso.sh
test/arch-arm-thumb-interwork.sh
test/pie.sh
test/z-nodump.sh
test/ar-alignment.sh
test/range-extension-thunk2.sh
test/defsym-lto.sh
test/z-dynamic-undefined-weak.sh
test/sysroot-linker-script.sh
test/lto-archive3.sh
test/ifunc-address-equality-exported.sh
test/range-extension-thunk4.sh
test/ifunc-address-equality.sh
test/init.sh
test/version-script21.sh
test/default-symver-version-script.sh
test/unresolved-symbols.sh
test/version-script23.sh
test/lto-version-script.sh
test/section-start.sh
test/reloc-rodata.sh
test/arch-x86_64-mergeable-strings.sh
test/visibility.sh
test/response-file.sh
test/ifunc-dynamic.sh
test/gnu-retain.sh
test/ifunc-alias.sh
test/large-alignment-dso.sh
test/exclude-libs2.sh
test/version-script10.sh
test/symbol-rank.sh
test/arch-x86_64-incompatible-libs-linker-script.sh
test/tls-alignment-multi.sh
test/depaudit.sh
test/version-script22.sh
test/common-ref.sh
test/relro.sh
test/interpose.sh
test/global-offset-table.sh
test/z-origin.sh
test/tls-ld.sh
test/arch-x86_64-apx-gotpcrelx.sh
test/exception.sh
test/initfirst.sh
test/arch-x86_64-incompatible-libs.sh
test/tls-large-alignment.sh
test/exclude-libs.sh
test/bsymbolic.sh
test/init-array-priorities.sh
test/debug-macro-section.sh
test/absolute-symbols.sh
test/ctors-in-init-array.sh
test/lto-comdat.sh
test/defsym2.sh
test/symbol-version2.sh
test/z-separate-code.sh
test/version-script8.sh
test/weak-undef4.sh
test/weak-undef-dso.sh
test/tls-pic.sh
test/ifunc-noplt.sh
test/bsymbolic-functions.sh
test/arch-x86_64-warn-execstack.sh
test/tls-gd-to-ie.sh
test/ifunc-funcptr.sh
test/linker-script3.sh
test/arch-x86_64-z-rewrite-endbr3.sh
test/dead-debug-sections.sh
test/version-script19.sh
test/tbss-only.sh
test/linker-script2.sh
test/shared.sh
test/tls-gd-noplt.sh
test/z-now.sh
test/gnu-hash.sh
test/weak-export-dso.sh
test/version-script11.sh
test/tail-call.sh
test/arch-x86_64-incompatible-libs2.sh
test/arch-x86_64-apx-tlsdesc.sh
test/version-script6.sh
test/copyrel-relro.sh
test/exclude-libs3.sh
test/dynamic-list.sh
test/hello-static.sh
test/arch-x86_64-apx-gottpoff.sh
test/tlsdesc.sh
test/arch-arm-tlsdesc.sh
test/arch-x86_64-large-bss.sh
test/arch-x86_64-exception-mcmodel-large.sh
test/verbose.sh
test/arch-riscv64-relax-align.sh
test/start-stop-symbol.sh
test/version-script7.sh
test/tls-df-static-tls.sh
test/ifunc-export.sh
test/arch-x86_64-z-shstk.sh
test/many-input-sections.sh
test/symbol-version3.sh
test/issue646.sh
test/protected-dynsym.sh
test/execstack.sh
test/fatal-warnings.sh
test/default-symver.sh
test/tls-nopic.sh
test/version-script20.sh
test/z-sectionheader.sh
test/z-stack-size.sh
test/linker-script6.sh
test/tls-ld-noplt.sh
test/large-max-page-size-strip.sh
test/emit-relocs.sh
test/version-script-search-paths.sh
test/compressed-debug-info.sh
test/arch-x86_64-init-array-readonly.sh
test/common-archive.sh
test/textrel.sh
test/version-script.sh
test/lto-archive2.sh
test/z-dynamic-undefined-weak2.sh
test/relax-got-load.sh
test/arch-x86_64-tls-gd-mcmodel-large.sh
test/z-max-page-size.sh
test/entry.sh
test/arch-x86_64-tlsdesc.sh
test/weak-export-dso2.sh
test/copyrel-relro2.sh
test/depaudit2.sh
test/weak-undef.sh
test/z-nodefaultlib.sh
test/version-script18.sh
test/arch-x86_64-apx-gottpoff2.sh
test/as-needed.sh
test/arch-x86_64-tls-module-base.sh
test/arch-x86_64-section-alignment.sh
test/symbol-version5.sh
test/arch-x86_64-preinit-array.sh
test/tls-le.sh
test/arch-x86_64-ifunc-alias.sh
test/linker-script-defsym.sh
```
</details>

All output of `cargo test` has been consolidated into [this](https://gist.github.com/lapla-cogito/fbe80dee61abe2f94d8e0cb51e66e875)

davidlattimore commented:
Thanks for working on this PR! Don't worry about those two test failures on ubuntu:25.10, they're not related to your change and have been happening all day. We might need to disable 25.10 if they continue.

My first question is how does this use wild as the linker when running the tests?

lapla-cogito commented:
@davidlattimore Sorry, I simply forgot to add the `PATH` modification and have now corrected it (I've also updated the list of passed tests and the output of `cargo test`).

davidlattimore commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2429,66 @@ fn integration_test(
+            let output = Command::new("bash")
+                .arg("-c")
+                .arg(entry.path())
+                .env("PATH", format!("{}:{path}", wild_dir.display()))
```
I thought that Mold's tests relied on the current working directory containing a binary called "ld". I guess if it can't find `ld` in the current directory, it would look on the path, so this might work, although I think currently it'd find the release build of wild. Wild's repository root contains a symlink called "ld", but it's a symlink to `target/release/wild` which wouldn't be suitable if the tests were being run without `--release`. The `fakes-debug` directory contains similar symlinks, but for debug builds.

One useful manual test for your tests is to introduce some bug - say change wild so that it always panics, then run `cargo test test_external_test_suits` and verify that all of the tests now fail.

marxin commented:
Have a couple of general comments:
1. I would really like direct integration of the individual tests from the Mold's test suite using the following compile-time glob listing: https://docs.rs/rstest/0.25.0/rstest/attr.rstest.html#files-path-as-input-arguments.
2. Having that, the test harness will run the tests in parallel, as we won't have to implement our own parallelism.
3. It would also be great to list all the known failing Mold tests and have `XFAIL` (fail the test if the test is actually passing).
4. As a follow-up, I know Mold's test suite supports running cross-arch tests using QEMU; it would be nice to support it.
5. Filtering of `tests/arch-*` should depend on the host system.
6. To prevent the Mold test-suite output, I would consider suppressing `set -x`, which is set by common.inc.

marxin commented:
One potential integration opportunity would be adding Mold as a submodule of the project.
@davidlattimore: What do you think about it?

davidlattimore commented:
> One potential integration opportunity would be adding Mold as a submodule of the project.
> @davidlattimore: What do you think about it?

I guess that should be OK. Perhaps in a subdirectory. e.g. `external_test_suites/mold`

lapla-cogito commented:
> > One potential integration opportunity would be adding Mold as a submodule of the project.
> > @davidlattimore: What do you think about it?
> 
> I guess that should be OK. Perhaps in a subdirectory. e.g. `external_test_suites/mold`

I'm generally reluctant to use external test suites as submodules:
1. Submodules require the entire repository. Since we only need the test suite itself, this is excessively redundant.
2. If there are updates to the external test suite and we want to incorporate those changes, we'd need to explicitly update the submodule (this is just one example - we could maintain the mold test suite itself as a GitHub Actions cache and periodically update the cache at fixed intervals like weekly via cron jobs, allowing CI to automatically obtain the latest results without manual intervention).

marxin commented:
> * Submodules require the entire repository. Since we only need the test suite itself, this is excessively redundant.

Well, yes, but we're speaking about a rather small git repository.

> * If there are updates to the external test suite and we want to incorporate those changes, we'd need to explicitly update the submodule (this is just one example - we could maintain the mold test suite itself as a GitHub Actions cache and periodically update the cache at fixed intervals like weekly via cron jobs, allowing CI to automatically obtain the latest results without manual intervention).

That's, from my perspective, a desired feature. Every time we pull Mold's changes, one needs to triage new tests and mark them as passing or failing. Otherwise, everyone from Wild's developers will stick to a different revision of Mold, and then we'll see test suite pass rate inconsistency.

Plus, having the test suite at a fixed location will simplify any current code that deals with the proper finding of the tests.

mati865 commented:
You also do not always have to initialize the submodules. This is how the Rust compiler repository works. It has some necessary submodules, like docs, but you can skip others depending on your build configuration.

davidlattimore commented:
Or, if we wanted it to be driven by `test-config.toml` then we could have a `build.rs` that parses `test-config.toml` and if the mold tests are enabled (which could be a boolean in the config) prints `cargo::rustc-cfg=mold_tests`. It'd also need to print `cargo::rustc-check-cfg=mold-tests` regardless of whether they're enabled. It should also print `cargo::rerun-if-changed=test-confiig.toml`. Then in the code, you can use `#[cfg(mold-tests)]`.

marxin commented:
I'm curious if we could simply remove the `build.rs` file (and the integration into `config.toml`) and manually drive the testing of Mold's test suite using a `cfg` feature? Doing that, we'll end up with a much smaller change?

marxin commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2435,43 @@ fn integration_test(
+#[rstest]
+#[cfg(mold_tests)]
+fn exec_mold_tests(
+    #[files("../external_test_suites/mold/test/*.sh")] mold_test: PathBuf,
```
To be honest, I am pretty excited about how elegant the listing is and how the tests run in parallel, great!
Are you planning to come up with a list of excluded tests via `#exclude()` directive:
https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments
?

lapla-cogito replied:
Yes, I agree that filtering tests using something like the `exclude` attribute would be necessary. However, this wasn't originally planned as part of this PR. For the future of this feature, current plan is to automatically categorize test execution results from mold tests based on error messages, aggregate them into CSV files, and integrate them into CI. Therefore, deciding what mold test cases should be excluded when implementing this categorization feature as a patch would be most effective when determining which test cases should be excluded.

lapla-cogito replied:
I've updated the list of currently passing tests the test execution results presented at the beginning of this PR. Based on this, I'm going to begin manual investigation.

marxin commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2435,43 @@ fn integration_test(
+    let output = Command::new("bash")
+        .arg("-c")
+        .arg(&mold_test)
+        .env("PATH", format!("{wild_dir:?}:{path}"))
```
I noticed many of the tests fail because of the following error:
```
+ ./mold --relocatable -o out/test/x86_64/many-input-sections2/b.o out/test/x86_64/many-input-sections2/a.o
/home/marxin/Programming/wild/external_test_suites/mold/test/many-input-sections2.sh: line 11: ./mold: No such file or directory
++ on_error 11
++ code=127
```

That means, we might consider adding a `mold` symlink to `fakes-debug` as well?

lapla-cogito replied:
Thank you! I hadn't noticed that.

davidlattimore commented:
> > I'm curious if we could simply remove the build.rs file (and the integration into config.toml) and manually drive the testing of Mold's test suite using a cfg feature? Doing that, we'll end up with a much smaller change?
> 
> I was exactly struggling with the best way to centrally manage `ExternalTestSuites` when your suggestion about using feature flags seemed particularly promising. Thanks!

I guess that means that we don't need anything in `test-config.toml` - whether we look for and run mold's test suite is controlled entire by `--features mold-tests` or something like that. That is nice and simple :) It also has the advantage that we can choose from run to run whether we want to run those tests.

davidlattimore commented on fakes/mold:
```diff
@@ -0,0 +1 @@
+fakes/ld
```
Can you try running `fakes/mold --version` and confirm that it works and reports that it's Wild? Just that this symlink target doesn't look right to me

davidlattimore commented on wild/Cargo.toml:
```diff
@@ -37,7 +37,13 @@ toml = { workspace = true }
 wait-timeout = { workspace = true }
 which = { workspace = true }
 
+[build-dependencies]
```
Looks like you missed these?

marxin commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2426,35 @@ fn integration_test(
+) -> Result {
+    let path = env::var("PATH").unwrap();
+    let current_dir = env::current_dir()?;
+    let wild_dir = current_dir.parent().unwrap().join("fakes-debug");
```
Seems to me, various tests that execute `./mold` fail:
```
Error: Mold test `/home/marxin/Programming/wild/external_test_suites/mold/test/relocatable-many-sections.sh` failed with status: exit status: 127
stdout: Testing relocatable-many-sections ... command failed: 19: ./mold -r -o $t/c.o $t/a.o $t/b.o

stderr: + nm mold
+ grep __tsan_init
nm: 'mold': No such file
+ seq 1 80000
+ sed 's/.*/.section .data.&,"aw"\n.word 0\n/g'
+ cc -c -xassembler -o out/test/x86_64/relocatable-many-sections/a.o -
+ cc -c -xc -o out/test/x86_64/relocatable-many-sections/b.o -
+ cat
+ ./mold -r -o out/test/x86_64/relocatable-many-sections/c.o out/test/x86_64/relocatable-many-sections/a.o out/test/x86_64/relocatable-many-sections/b.o
/home/marxin/Programming/wild/external_test_suites/mold/test/relocatable-many-sections.sh: line 19: ./mold: No such file or directory
++ on_error 19
++ code=127
++ echo 'command failed: 19: ./mold -r -o $t/c.o $t/a.o $t/b.o'
++ trap - EXIT
++ exit 127
```

As seen, many tests expect to be executed from a location where `./mold` is present. That said, I would probably change working directory to the root of the tests, before running the script.

marxin commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2426,35 @@ fn integration_test(
+fn exec_mold_tests(
+    #[files("../external_test_suites/mold/test/*.sh")] mold_test: PathBuf,
+) -> Result {
+    let path = env::var("PATH").unwrap();
```
I know it's very unlikely the `env` is not set, but I would rather add `.context` and use `?` (instead of the unwrap);

lapla-cogito commented on .github/workflows/ci.yml:
```diff
@@ -59,17 +59,18 @@ jobs:
+      - run: zypper in -y gcc gcc-c++ glibc-devel-static clang lld curl rustup bubblewrap git
         if: ${{ contains(matrix.container, 'opensuse') }}
-      - run: apk add build-base lld clang bash curl
+      - run: apk add build-base lld clang bash curl git
```
While `actions/checkout@v4` itself can function without installing a new git binary, it appears to use the installed git version, which is apparently outdated. The workflow attempts to update repositories via the GitHub REST API, which seems to be causing the failure ([example here](https://github.com/davidlattimore/wild/actions/runs/15852585930/job/44689652692)). Therefore, we explicitly install git.

marxin replied:
Thanks for the explanation!

marxin commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2426,35 @@ fn integration_test(
+    let current_dir = env::current_dir()?;
+    let wild_dir = current_dir.parent().unwrap().join("fakes-debug");
+
+    let output = Command::new("bash")
```
One last nit: can you please combine stdout+stderr directly into one stream? Having that, we'll better see the interleaved messages like Wild's warnings directly at the place they happen.

lapla-cogito replied:
Right, that does sound more convenient. Fixed, thanks!

davidlattimore commented on wild/tests/integration_tests.rs:
```diff
@@ -2425,6 +2426,35 @@ fn integration_test(
+    let output = Command::new("bash")
+        .arg("-c")
+        .arg(&mold_test)
+        .env("PATH", format!("{wild_dir:?}:{path}"))
```
Cargo will build binaries before it runs integration tests. You can find the location of those binaries at build time via the environment variable `CARGO_BIN_EXE_<name>` e.g. `env!("CARGO_BIN_EXE_wild")`. So for maximum correctness, the test would need to create a temporary directory and put symlinks into it that pointed to the actual location of the wild binary. If would be somewhat inconvenient however if that temporary directory got deleted after the tests finished, because it'd make it harder to do manual checks of what happened - e.g. the out directory would vanish. I suppose one option would be to always use the same directory and update the symlinks when the test runs if they're not correct, then .gitignore the whole directory.

davidlattimore replied:
But I think using `fakes-debug` should work OK so long as you don't `cargo test --release`, specify some other profile or override the target directory.
